File: server/internal/repository/category.go
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"cashcompass-server/internal/model"
)

type CategoryRepository interface {
	List(ctx context.Context) ([]model.Category, error)
	GetByID(ctx context.Context, id int) (model.Category, error)
	Create(ctx context.Context, label string) (model.Category, error)
	UpdateLabel(ctx context.Context, id int, label string) error
	Delete(ctx context.Context, id int) error
}

type SQLiteCategoryRepository struct {
	db *sql.DB
}

func NewSQLiteCategoryRepository(db *sql.DB) *SQLiteCategoryRepository {
	return &SQLiteCategoryRepository{db: db}
}

const categoryColumns = "id, label, created_at, updated_at"

func scanCategory(s rowScanner) (model.Category, error) {
	var c model.Category
	var createdAt, updatedAt string
	if err := s.Scan(&c.ID, &c.Label, &createdAt, &updatedAt); err != nil {
		return model.Category{}, err
	}
	c.CreatedAt, _ = time.Parse("2006-01-02 15:04:05", createdAt)
	c.UpdatedAt, _ = time.Parse("2006-01-02 15:04:05", updatedAt)
	return c, nil
}

func (r *SQLiteCategoryRepository) List(ctx context.Context) ([]model.Category, error) {
	rows, err := r.db.QueryContext(ctx,
		"SELECT "+categoryColumns+" FROM categories ORDER BY label ASC")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var categories []model.Category
	for rows.Next() {
		c, err := scanCategory(rows)
		if err != nil {
			return nil, err
		}
		categories = append(categories, c)
	}
	return categories, rows.Err()
}

func (r *SQLiteCategoryRepository) GetByID(ctx context.Context, id int) (model.Category, error) {
	row := r.db.QueryRowContext(ctx,
		"SELECT "+categoryColumns+" FROM categories WHERE id = ?", id)
	c, err := scanCategory(row)
	if err == sql.ErrNoRows {
		return model.Category{}, fmt.Errorf("category %d not found", id)
	}
	return c, err
}

func (r *SQLiteCategoryRepository) Create(ctx context.Context, label string) (model.Category, error) {
	res, err := r.db.ExecContext(ctx,
		"INSERT INTO categories (label) VALUES (?)", label)
	if err != nil {
		return model.Category{}, err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return model.Category{}, err
	}
	return r.GetByID(ctx, int(id))
}

func (r *SQLiteCategoryRepository) UpdateLabel(ctx context.Context, id int, label string) error {
	_, err := r.db.ExecContext(ctx,
		"UPDATE categories SET label = ?, updated_at = datetime('now') WHERE id = ?", label, id)
	return err
}

func (r *SQLiteCategoryRepository) Delete(ctx context.Context, id int) error {
	_, err := r.db.ExecContext(ctx, "DELETE FROM categories WHERE id = ?", id)
	return err
}

File: src/repository/category.py
from __future__ import annotations
from src.db import Database
from src.models import Category
from src.utils.ids import generate_uid


def _row_to_category(row) -> Category:
    return Category(
        id=row["id"],
        uid=row["uid"],
        label=row["label"],
        created_at=row["created_at"] or "",
        updated_at=row["updated_at"] or "",
    )


class CategoryRepository:
    def __init__(self, db: Database):
        self.db = db

    def list(self) -> list[Category]:
        rows = self.db.execute(
            "SELECT id, uid, label, created_at, updated_at FROM categories ORDER BY label ASC"
        ).fetchall()
        return [_row_to_category(r) for r in rows]

    def get_by_id(self, id: int) -> Category:
        row = self.db.execute(
            "SELECT id, uid, label, created_at, updated_at FROM categories WHERE id = ?", [id]
        ).fetchone()
        if row is None:
            raise ValueError(f"category {id} not found")
        return _row_to_category(row)

    def create(self, label: str) -> Category:
        cur = self.db.execute("INSERT INTO categories (uid, label) VALUES (?, ?)", [generate_uid("cat"), label])
        self.db.commit()
        return self.get_by_id(cur.lastrowid)

    def update_label(self, id: int, label: str):
        self.db.execute(
            "UPDATE categories SET label = ?, updated_at = datetime('now') WHERE id = ?", [label, id]
        )
        self.db.commit()

    def delete(self, id: int):
        self.db.execute("DELETE FROM categories WHERE id = ?", [id])
        self.db.commit()
